### PR TITLE
Fixes 9880

### LINKED ISSUE #9880: `index out of bounds` errors should show `index` and `bound`
`index out of bounds` errors should show `index` and `bound` in error message; it would often save a lot of debugging time

the code needs to be updated in a number of places:
```
compiler/ccgexprs.nim:878:39:        localError(p.config, x.info, "index out of bounds")
compiler/semfold.nim:492:37:      localError(g.config, n.info, "index out of bounds: " & $n)
compiler/semfold.nim:496:41:    else: localError(g.config, n.info, "index out of bounds: " & $n)
compiler/semfold.nim:504:37:      localError(g.config, n.info, "index out of bounds: " & $n)
compiler/vm.nim:424:26:  errIndexOutOfBounds = "index out of bounds"
lib/core/typeinfo.nim:197:39:      raise newException(IndexError, "index out of bounds")
lib/core/typeinfo.nim:204:39:      raise newException(IndexError, "index out of bounds")
lib/core/typeinfo.nim:214:39:      raise newException(IndexError, "index out of bounds")
lib/core/typeinfo.nim:222:39:      raise newException(IndexError, "index out of bounds")
lib/pure/base64.nim:71:13:    # avoid index out of bounds when lineLen == encoded length
lib/pure/collections/sharedstrings.nim:52:41:  else: raise newException(IndexError, "index out of bounds")
lib/pure/collections/sharedstrings.nim:56:41:  else: raise newException(IndexError, "index out of bounds")
lib/pure/subexes.nim:88:44:  if result >=% a.len: raiseInvalidFormat("index out of bounds: " & $result)
lib/system/chcks.nim:19:25:  sysFatal(IndexError, "index out of bounds")
lib/system/jssys.nim:161:35:  raise newException(IndexError, "index out of bounds")
```

i suggest using a single routine to generate the error message, eg:
```nim
template formatErrorIndexBound(index: int, n: int): string =
  "index out of bounds: index: " & $index &" n: " & $n
```

### PR DESCRIPTION
**Pull request by BontaVlad**
Fixes #9880 
I'm still learning nim, of course this PR needs adult supervision.

### CODE REVIEW

**Comment by krux02** [comment]
Can you add a test case please? Apart from that, I agree, the error should contain the bounds.

**Comment by narimiran** [comment]
> I'm still learning nim, of course this PR needs adult supervision.

Your contributions (I'm including here your comment on #10198, which led to a pull-request and a fix) are very appreciated and welcome. Keep them coming as you learn Nim.

**Comment by narimiran** [comment]
From my understanding (and I might be wrong — please correct me), `formatErrorIndexBound` returns a valid index range (usually from 0 to len-1), which would mean that in several places where you pass `someLength+1` you should pass `someLength-1`.

(Don't correct anything until somebody confirms this :))

**Comment by krux02** [comment]
@narimiran A test case that triggers these exceptions should clear that question up.

**Comment by BontaVlad** [comment]
I'm on board with having tests, I just need some directions on how do I trigger these exceptions and also how to cache the mentioned exceptions.

**Comment by krux02** [comment]
you should be able to raise the exception when you acces an arrar or a seq outside of its bounds. With try catch you can catch the exception and write an assert.

File: lib/system/jssys.nim [hunk 1]
```diff
@@ -157,8 +159,8 @@ proc raiseDivByZero {.exportc: "raiseDivByZero", noreturn, compilerProc.} =
 proc raiseRangeError() {.compilerproc, noreturn.} =
   raise newException(RangeError, "value out of range")
 
-proc raiseIndexError() {.compilerproc, noreturn.} =
-  raise newException(IndexError, "index out of bounds")
+proc raiseIndexError[T](i, a, b: T) {.compilerproc, noreturn.} =
```
[the remark this answers is not in the document]
Comment (timotheecour) [reply lib/system/jssys.nim]: should'nt that give CT error then?

[the remark this answers is not in the document]
Comment (Araq) [reply lib/system/jssys.nim]: You are supposed to know what you're doing when touching system.nim.

**Comment by BontaVlad** [comment]
What do I need to do on this PR to get it merged?

**Comment by Araq** [comment]
Don't use a generic compilerproc, follow what the compilerproc for the C backend does.

File: lib/system/jssys.nim (hunk 1 above)
[the remark this answers is not in the document]
Comment (timotheecour) [reply lib/system/jssys.nim]: filed https://github.com/nim-lang/Nim/issues/10516 @BontaVlad that should explain the problem:
you'll only get an error when multiple types are instantiated
* do you actually need genericity here or can arguments `foo` be explicitly converted to `int` using `int(foo)` ?
* else, try @araq's solution from https://github.com/nim-lang/Nim/pull/10228#issuecomment-459324981

[the remark this answers is not in the document]
Comment (timotheecour) [reply lib/system/jssys.nim]: > * do you actually need genericity here or can arguments `foo` be explicitly converted to `int` using `int(foo)` ?

ok looks like ur last commit does that, marking as resolved

**Comment by BontaVlad** [comment]
Since #10610 was merged, can I close this PR?

**Comment by timotheecour** [comment]
@BontaVlad yes, I don't think there's any remaining items left out

**Pull request closed without merging**